src/librustc/ty/util.rs

```diff
@@ -525,6 +393,142 @@ impl<'a, 'gcx, 'tcx> TyCtxt<'a, 'gcx, 'tcx> {
     }
 }
 
+struct TypeIdHasher<'a, 'gcx: 'a+'tcx, 'tcx: 'a> {
+    tcx: TyCtxt<'a, 'gcx, 'tcx>,
+    svh: &'a Svh,
+    state: SipHasher
+}
+
+impl<'a, 'gcx, 'tcx> TypeIdHasher<'a, 'gcx, 'tcx> {
+    fn hash<T: Hash>(&mut self, x: T) {
+        x.hash(&mut self.state);
+    }
+
+    fn hash_discriminant_u8<T>(&mut self, x: &T) {
+        let v = unsafe {
+            intrinsics::discriminant_value(x)
+        };
+        let b = v as u8;
+        assert_eq!(v, b as u64);
+        self.hash(b)
+    }
+
+    fn def_id(&mut self, did: DefId) {
+        let h = if did.is_local() {
+            self.svh.clone()
+        } else {
+            self.tcx.sess.cstore.crate_hash(did.krate)
+        };
+        self.hash(h);
+        self.hash(did.index);
```

Comment:
Def-ids are not stable across compilations. It would be better to hash the (stringified) def-path, which is stable. I could imagine this causing some subtle problems with incremental compilation, actually, where a def-id of some struct changes but the struct itself does not. Unfortunately, some of the helpers needed to conveniently hash the def-path are not landed yet (they are in an upcoming PR). At minimum we should open a FIXME here.